### PR TITLE
support `GC.stat(:time)` with gc_enter/gc_exit

### PR DESCRIPTION
**Pull request by ko1**
https://bugs.ruby-lang.org/issues/10917

    GC measurement feature

    * `GC.measure_total_time = true` enables total time measurement (default: true)
    * `GC.measure_total_time` returns current flag.
    * `GC.total_time` returns measured total time in nano seconds.
    * `GC.stat(:time)` (and Hash) returns measured total time in milli seconds.

### CODE REVIEW

File: gc.c [hunk 1]
```diff
@@ -798,6 +798,8 @@ typedef struct rb_objspace {
 	size_t total_freed_objects;
 	size_t total_allocated_pages;
 	size_t total_freed_pages;
+        size_t total_time_ns;
```
Comment (byroot) [inline gc.c:801]: Not sure how important is is, but if I'm not mistaken storing nanoseconds in `size_t` means that on 32bit systems the counter will overflow after 4.3 seconds spent in GC.

**Comment by casperisfine** [comment]
@ko1 is there anything I could do to help this move forward? Beside the 32 bits issue this looks good no?

**Comment by djudd** [comment]
I want to +1 this (& had submitting a similar PR on my personal backlog).

At Stripe we've run into not-infrequent cases in production where the incremental sweep time that's not captured by GC::Profiler dominates total GC time, and relying on GC::Profiler misled us for a long while. We currently run a C extension that uses `RB_INTERNAL_EVENT_GC_ENTER` and `_EXIT` hooks like [gctrack](https://github.com/Shopify/gctrack/blob/master/ext/gctrack/gctrack.c), instead, and found that it had comparable (~1%) overhead to enabling GC::Profiler, which we now don't. I imagine this PR's approach has lower overhead, and it would also avoid misleading people. (I'd actually suggest backporting this to older Rubies and updating or removing `GC::Profiler.total_time` too, since I think the error is big enough to qualify as a bug.)

Here's an example script that, for me, reproduces a large undercount of GC time by GC::Profiler, with Ruby 2.7 defaults:

```ruby
require 'gctrack'

def make_some_garbage
  1_000_000.times.map {Class.new}
end

GC::Profiler.enable
GC::Tracker.enable

persistent_heap = make_some_garbage

profiler_start = GC::Profiler.total_time
GC::Tracker.start_record

make_some_garbage

tracker_total = GC::Tracker.end_record.last
profiler_total = GC::Profiler.total_time - profiler_start

puts "GC::Tracker says: #{(tracker_total / 1_000_000_000.0).round(3)}s, GC::Profiler says: #{profiler_total.round(3)}s"
```

Example output: `GC::Tracker says: 0.494s, GC::Profiler says: 0.142s`

**Comment by djudd** [comment]
Like @casperisfine I'm happy to help move this forward myself if I can. 

It might make sense to steal portable timing code from Stackprof, or if not to default `MEASURE_GC` to `_POSIX_MONOTONIC_CLOCK`: https://github.com/tmm1/stackprof/blob/master/ext/stackprof/stackprof.c#L43-L74

**Comment by casperisfine** [comment]
I added this back to the last developer meeting an @ko1 confirmed this will make it for 3.1: https://github.com/ruby/dev-meeting-log/blob/master/DevelopersMeeting20211021Japan.md#feature-10917-add-gcstattotal_time-when-gc-profiling-enabled-byroot

> I'd actually suggest backporting this to older Rubies and updating or removing GC::Profiler.total_time too

That's way outside the backport policy.

**Comment by ko1** [comment]
I cleanup the patch and add some methods:

```
    GC measurement feature

    * `GC.measure_total_time = true` enables total time measurement (default: true)
    * `GC.measure_total_time` returns current flag.
    * `GC.total_time` returns measured total time in nano seconds.
    * `GC.stat(:time)` (and Hash) returns measured total time in milli seconds.
```

**Comment by byroot** [comment]
I think we should add documentation for the `time` key in `GC.stat`, and maybe also mention it will be zero if measure isn't enabled.

**Comment by ko1** [comment]
Another points are

    * `GC.total_time` returns measured total time in nano seconds.
    * `GC.stat(:time)` (and Hash) returns measured total time in milli seconds.

ns and ms. it should be discussed in redmine ticket.

**Comment by byroot** [comment]
What about `GC.total_time` returns in ms but as a float? So that the scale is the same, less surprising, but you still get the higher precision (even though I doubt `ns` is really useful).

**Comment by ko1** [comment]
> What about `GC.total_time` returns in ms but as a float? So that the scale is the same, less surprising, but you still get the higher precision (even though I doubt `ns` is really useful).

If we allow the Float, I vote to return seconds in Float because `GC::Profiler.total_time` returns it. But it is a bit slow than integer... nobody care!!

**Pull request merged**